x/sync/syncmanager.go

```diff
@@ -344,14 +344,29 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
 	m.completeWorkItem(ctx, workItem, largestHandledKey, rootID, proof.EndProof)
 }
 
-// findNextKey attempts to find the first key larger than the lastReceivedKey that is different in the local merkle vs the merkle that is being synced.
-// Returns the first key with a difference in the range (lastReceivedKey, rangeEnd), or nil if no difference was found in the range
+// findNextKey returns the start of the key range that should be fetched next.
+// Returns nil if there are no more keys to fetch  up to [rangeEnd].
+//
+// If the last proof received contained at least one key-value pair, then
+// [lastKeyReceived] is the greatest key in the key-value pairs received.
+// Otherwise it's the end of the range for the last proof received.
+//
+// [rangeEnd] is the end of the range that we want to fetch.
+//
+// [endProof] is the end proof of the last proof received.
+// Namely it's a proof for [lastKeyReceived].
```

Comment:
It's also possible it's an exclusion proof if the next key is outside of rangeEnd